Change mainline advancement modification

Pentarctagon opened the pull request:
The two commits explain their own reasoning.  Beyond that, the remaining question would be whether an advancement planning mod would:
* be a mainlining of the existing Plan Unit Advance modification; or
* be something brand new

The first option also of course depends on whether @vgaming is fine with their modification being mainlined or not.

vgaming commented:
Hi, do I understand it correctly that we're considering the two commits above, PLUS adding some Lua/CPP source code to allow pre-selecting advances if they happen on enemy turn?

Depending on that, and having maintained the current "Plan Unit Advance" add-on, I could give some implementation/feedback details further.

Overall, I'm fine if wesnoth will make this feature internal (and thus make PUA obsolete). Removing the clutter and making people's playing experience smoother is definitely good by me :)

soliton- commented:
If you want to change advancement to default to the first one I would suggest to make sure that the first one is the most sensible choice.

Pentarctagon commented:
@vgaming What's being considered is either: (a) your modification as-is; or (b) a separate modification.  Implementing it in C++ wasn't something I was considering.

@soliton- I'll take a look.

vgaming commented:
@soliton- oh right, that makes a lot of sense! That's how the add-on should then work, I think.

@Pentarctagon All clear, all good :+1: 

Unrelated: not sure if we should randomize the advancement options for units at startup then :thinking: I guess we can return to this question when implementing the Lua part.

Pentarctagon commented:
> Unrelated: not sure if we should randomize the advancement options for units at startup then thinking I guess we can return to this question when implementing the Lua part.

That would effectively negate the change to not randomly choose an advancement - choosing a random advancement from a predefined list is the same as always choosing the first advancement from a randomly ordered list.

vgaming commented:
@Pentarctagon not exactly. First, an add-on can then conveniently set a preference for advancing if it happens on enemy turn while not preventing the user to make a choice if advance happens on their turn. Second, the randomization code can be moved from C++ to Lua, which I believe is a good thing because the CPP core can be more lightweight (hopefully more maintainable?). Third, but that is really minor, randomization at advance time is slightly different than randomization at recruit time if you reload your game (but again that can be disregarded).

vgaming commented:
@Pentarctagon and generally, this question is for later when if/after we accept the PR. It can obviously go both ways, the current PR doesn't prevent any future decision here.

Pentarctagon commented:
In any case I suppose, running `grep -Iirn "advances_to=.*,.*"` gives the results:
```
data/campaigns/Secrets_of_the_Ancients/units/Delinquent.cfg:15:              advances_to=Footpad,SotA Dark Adept
data/campaigns/Secrets_of_the_Ancients/units/Ruffian.cfg:8:                  advances_to=Footpad,Thug,SotA Dark Adept
data/campaigns/Under_the_Burning_Suns/units/undead/Ethereal_Ghost.cfg:15:    advances_to=EWraith,EShadow
data/campaigns/Under_the_Burning_Suns/units/quenoth/Fighter.cfg:20:          advances_to=Quenoth Warrior,Quenoth Flanker
data/campaigns/Under_the_Burning_Suns/units/quenoth/Scout.cfg:15:            advances_to=Quenoth Pathfinder,Quenoth Archer
data/campaigns/Under_the_Burning_Suns/units/quenoth/Mystic.cfg:16:           advances_to=Quenoth Shaman,Quenoth Sun Singer
data/campaigns/Under_the_Burning_Suns/units/quenoth/Tauroch_Rider.cfg:24:    advances_to=Tauroch Vanguard,Tauroch Stalwart
data/campaigns/Liberty/units/Villagers.cfg:182:                              advances_to=Huntsman_Peasant,Ranger
data/campaigns/Dead_Water/units/Citizen.cfg:14:                              advances_to=Merman Brawler,Merman Fighter,Merman Hunter
data/core/units/dunefolk/Rider.cfg:17:                                       advances_to=Dune Raider,Dune Horse Archer,Dune Sunderer
data/core/units/dunefolk/Herbalist.cfg:22:                                   advances_to=Dune Apothecary,Dune Alchemist
data/core/units/dunefolk/Skirmisher.cfg:18:                                  advances_to=Dune Strider, Dune Falconer
data/core/units/dunefolk/Soldier.cfg:17:                                     advances_to=Dune Swordsman,Dune Captain,Dune Spearguard
data/core/units/merfolk/Hunter.cfg:15:                                       advances_to=Merman Spearman,Merman Netcaster
data/core/units/merfolk/Warrior.cfg:14:                                      advances_to=Merman Triton,Merman Hoplite
data/core/units/merfolk/Initiate.cfg:15:                                     advances_to=Mermaid Priestess,Mermaid Enchantress
data/core/units/undead/Spirit_Ghost.cfg:14:                                  advances_to=Wraith,Shadow
data/core/units/undead/Skeleton.cfg:14:                                      advances_to=Revenant,Deathblade
data/core/units/undead/Necro_Dark_Sorcerer.cfg:15:                           advances_to=Lich,Necromancer
data/core/units/elves/Fighter.cfg:15:                                        advances_to=Elvish Captain,Elvish Hero
data/core/units/elves/Archer.cfg:16:                                         advances_to=Elvish Ranger,Elvish Marksman
data/core/units/elves/Shaman.cfg:16:                                         advances_to=Elvish Druid,Elvish Sorceress
data/core/units/nagas/Dirkfang.cfg:15:                                       advances_to=Naga Ophidian,Naga Ringcaster
data/core/units/humans/Mage.cfg:16:                                          advances_to=White Mage,Red Mage
data/core/units/humans/Horseman.cfg:14:                                      advances_to=Knight,Lancer
data/core/units/humans/Mage_Red.cfg:15:                                      advances_to=Arch Mage,Silver Mage
data/core/units/humans/Peasant.cfg:15:                                       advances_to=Spearman,Bowman
data/core/units/humans/Woodsman.cfg:15:                                      advances_to=Bowman,Poacher
data/core/units/humans/Outlaw_Ruffian.cfg:14:                                advances_to=Thug,Footpad
data/core/units/humans/Horse_Knight.cfg:15:                                  advances_to=Paladin,Grand Knight
data/core/units/humans/Woodsman_Trapper.cfg:14:                              advances_to=Huntsman, Ranger
data/core/units/humans/Loyalist_Spearman.cfg:15:                             advances_to=Swordsman,Pikeman,Javelineer
data/core/units/drakes/Burner.cfg:14:                                        advances_to=Fire Drake,Drake Flare
data/core/units/drakes/Clasher.cfg:13:                                       advances_to=Drake Thrasher,Drake Arbiter
data/core/units/trolls/Whelp.cfg:14:                                         advances_to=Troll,Troll Rocklobber
data/core/units/saurians/Augur.cfg:14:                                       advances_to=Saurian Oracle,Saurian Soothsayer
data/core/units/goblins/Wolf_Rider.cfg:27:                                   advances_to=Goblin Knight,Goblin Pillager
data/core/units/goblins/Spearman.cfg:14:                                     advances_to=Goblin Impaler,Goblin Rouser
```
I'll take a closer look today or tomorrow when I have time, but from glancing through the list there's nothing that currently stands out - the two I would have changed being the Whelp (if the Troll Rocklobber had been first) and the Spearman (if the Javelineer had been first), but those are already not the case.
Also @Hejnewar do you have any thoughts/opinions on this?

Hejnewar commented:
Not really, currently as I recall advancements are in the same order that they are in unit file and that is also good because it's an order that people are used to. At the same time PUA is clear enough that changing order shouldnt affect it too negativley.

Pentarctagon commented:
@soliton- The current advancements list looks good to me.

Pentarctagon commented:
@vgaming What's the history/purpose behind:
```
local function array_to_set(arr)
	local result = {}
	for _, v in ipairs(arr) do
		result[v] = true
	end
	return result
end

local function array_filter(arr, func)
	local result = {}
	for _, v in ipairs(arr) do
		if func(v) then
			result[#result + 1] = v
		end
	end
	return result
end

--- works as anti-cheat and fixes tricky bugs in [male]/[female]/undead variation overrides
local function filter_overrides(unit, overrides)
	local possible_advances_array = original_advances(unit)
	local possible_advances = array_to_set(possible_advances_array)
	local filtered = array_filter(overrides, function(e) return possible_advances[e] end)
	return #filtered > 0 and filtered or possible_advances_array
end
```
What sort of bugs were there and how does it work as an anti-cheat?

vgaming commented:
@Pentarctagon hi,

I don't remember about the undead bug, should've written it down. Undead was the key how I discovered the bug IIRC. A way to re-find the bug could be to try level zombies of different variations without this fix. One possible idea of the reason: maybe PUA would otherwise set advancements for your, say, "saurian zombie" upgrades to the one you specified for "human zombies".

I remember about anti-cheat. The reason for this check is how `overrides` type of data is constructed. It originates from a `synchronize_choice` function call here: https://github.com/vgaming/pick_advance/blob/master/pick_advance/lua/main.lua#L136
Now what if the user has overridden the contents of `synchronize_choice` locally, e.g. to set their units to advance to lvl5 dragon? Because it's a local choice, players would just get a String denoting the desired advance back (lvl5 dragon in this example). This would result in any unit being able to advance to any other.

I do realize that using this cheat would require:
* a software developer, because you have to read the code and understand these details
* knowing that the fact of cheating would be immediately obvious to any observer

Still, in such case, I wanted to prevent a bad image for wesnoth where a hacker could do literally everything.

vgaming commented:
Function `array_to_set` transforms a table like this:
`{ [1] = "a", [2] = "b", [3] = "c"}` (which could also be written in Lua as `{"a", "b", "c"}`)
into
`{ ["a"] = true, ["b"] = true, ["c"] = true }`

Function `array_filter` filters an array of values by using a function (predicate). For example, given an array
`arr = {1, 2, 3, 4, 5}`
and a function `func = function(x) return x > 2  end`
calling `array_filter(arr, func)` should give the result `{3, 4, 5}`.

Note that without syntactic sugar, array `{1, 2, 3, 4, 5}` is `{[1] = 1, [2] = 2, [3] = 3, [4] = 4, [5] = 5}` and array `{3, 4, 5}` is `{ [1] = 3, [2] = 4, [3] = 5}`.

Pentarctagon commented:
Alright, thanks.

And for the `on_event`s, what was the reason for setting their priority to -91?

vgaming commented:
@Pentarctagon I think I wanted to execute this piece of code early (earlier than many other add-ons code) https://wiki.wesnoth.org/LuaWML:Events#on_event.lua

I didn't make any tests suggesting that need. I'd consider removing it. Can do that in the already existing PUA to give it some time for feedback BTW. Done: https://github.com/vgaming/pick_advance/commit/41e7fffbfbdba1e04d003114fe56fe7d6b10e4a7

Pentarctagon commented:
I'm not really bothered whether they're there or not; I was just curious if it had been needed to solve some issue that had come up previously.

Pentarctagon commented:
Remaining questions:
* @vgaming I added a bunch of comments, so can you check and make sure they're correct?
* @CelticMinstrel Would the json_format functions be useful as a general utility rather than specific to this modification?
* @Vultraz Would you be able to review the UI of the advancement choice dialog?
---
Also, as far as translations go, how would that be handled?  There are quite a few places where string concatenation is done (though mostly involving newlines to increase button height), and otherwise there are two strings that are (I assume) able to be translated by the definition and use of `local translate = wesnoth.textdomain "wesnoth"` but otherwise no strings displayed on the UI are preceded by an underscore.

vgaming commented on data/modifications/pick_advance/dialog.lua:
```diff
@@ -0,0 +1,168 @@
+
+
+function pickadvance.show_dialog_unsynchronized(advance_info, unit)
+-- dialog exit codes --
+  local help_code = -4
+  local reset_code = -3
+  local single_unit_code = -1
+  local all_units_code = 1
+--    
```
Nice idea to document dialog exit codes on top!
On a practical side, should we use tabs/spaces uniformly? These 4 lines are written with two spaces, below everything is in tabs.

Pentarctagon replied:
Thanks, fixed - by default my text editor is set to use spaces instead.

vgaming commented on data/modifications/pick_advance/dialog.lua:
```diff
@@ -0,0 +1,168 @@
+	local label = "Plan advance:"
+
+	local unit_type_options = advance_info.type_advances
+	--print_as_json("advances for", unit.type, unit_type_options)
```
TODO: delete this if `print_as_json` won't go into mainline

vgaming commented on data/modifications/pick_advance/dialog.lua:
```diff
@@ -0,0 +1,168 @@
+	for _, ut in ipairs(unit_type_options) do
+		options[#options + 1] = wesnoth.unit_types[ut]
+	end
+	local show_images = true
```
Not sure what I intended to do here, but I never finished really. Let's remove this constant and its one usage below

Pentarctagon replied:
Done.

vgaming commented on data/modifications/pick_advance/dialog.lua:
```diff
@@ -0,0 +1,168 @@
+		T.column { T.label { use_markup = true, label = label } },
+	}
+
+	local list_sub_row
```
With `show_images` inlined (see above), this would also be simplified.

Pentarctagon replied:
Done.

vgaming commented on data/modifications/pick_advance/dialog.lua:
```diff
@@ -0,0 +1,168 @@
+	local reset_button = T.button {
+		return_value = reset_code,
+		label = "\n" .. translate("Reset") .. "\n",
+		tooltip = "Reset advancements to default"
```
I wonder if the quality of the addon is sufficient for Wesnoth's quality standards. I mean, this text is untranslated. I never figured out how to do that properly and efficiently. I think adding the addon to mainline would still be a clear improvement over the current situation, just making the choice explicit here.

vgaming replied:
Oh sorry, I now see that this question was raised by @Pentarctagon already. Resolving it here to de-dup.

vgaming commented on data/modifications/pick_advance/dialog.lua:
```diff
@@ -0,0 +1,168 @@
+
+		is_game_override = is_reset or game_scope,
+		game_override = game_scope and options[item_result].id or nil,
+	}
```
! Important !
The current add-on was written with the intention of shrinking the possible advancement options into one (because otherwise wesnoth would randomize).

With the current PR (https://github.com/wesnoth/wesnoth/pull/5491), we could change the logic to *reorder* the possible advancement list so that user's choice comes first. The user would then still be able to manually pick an advance if it happens at their turn.

I've received comments from people that asked for the add-on to work this way already. I could try help doing this rewrite.

Pentarctagon replied:
That sounds like a good idea to me.

Pentarctagon replied:
@vgaming to clarify - was this something you were planning to look into in the near future (within the next week or so)?  Either way is fine - there's no reason a second PR can't be made for updating the implementation - but if not I'm probably going to go ahead and merge this with the current implementation.

vgaming replied:
Hey, I'm struggling between really wanting to do it, and not having time for it at the moment.

To explain the scope at least:
* We need to make the add-on reorder the advancement list, not shrinking it. (Q: should we re-check if other addons added new stuff to units?)
* We probably could eliminate or change the menu options. Right now the possibilities/options are: "assign for this unit", "assign for this unit and all new recruits". With the deterministic changes in mind, however, this could always default to changing recruits as well. It could be the worded as something like "pick advance for Mage" and it would assign advance for all Mages unassigned yet, plus all new Mage recruits.

vgaming replied:
Leaving it at the discretion of the PR author to change it or define the deadline. Both make sense.

Pentarctagon replied:
I think I'd favor going ahead and merging it as-is so it'd be available for 1.15.10's release; changing the implementation to make use of reordering the advancements list instead would definitely be nice to have, but it's also a lower priority than getting the modification mainlined.

> should we re-check if other addons added new stuff to units?

I don't think it'd be necessary, unless you have an idea for it?

> We probably could eliminate or change the menu options. Right now the possibilities/options are: "assign for this unit", "assign for this unit and all new recruits". With the deterministic changes in mind, however, this could always default to changing recruits as well. It could be the worded as something like "pick advance for Mage" and it would assign advance for all Mages unassigned yet, plus all new Mage recruits.

I think that changing the advancement for just a single unit is probably something people would want to do.  Assigning it to every unit of that type where they haven't made an explicit choice would also somewhat go against one of the reasons for making advancements non-random: if they like the default, they don't need to change the advancement.

Pentarctagon commented:
Updated to add translation support - @sevu could you double check whether this is correct?

Also, I've removed the help message since the same information is now shown on the tooltip when hovering over the respective buttons.

Pentarctagon commented:
The table-to-json functionality has been moved to mainline and renamed to `wesnoth.as_text(...)`.  A unit test has also been added to validate its output.

sevu commented on data/modifications/pick_advance/dialog.lua:
```diff
@@ -0,0 +1,131 @@
+-- << pickadvance_dialog
+
+-- #textdomain wesnoth
```
Not needed, line 7 is enough.

Pentarctagon replied:
Updated.

sevu commented on data/lua/as_text.lua:
```diff
@@ -0,0 +1,66 @@
+		buffer[#buffer + 1] = '"' .. escape(obj) .. '"'
+	elseif _type == "boolean" or _type == "number" then
+		buffer[#buffer + 1] = tostring(obj)
+	elseif _type == "userdata" then
```
Should there be a case for translatable strings in the list? I think they had their own type or they are userdata.

Pentarctagon replied:
According to [here](https://wiki.wesnoth.org/LuaAPI#Wesnoth_Types) they'd be userdata.

sevu commented on data/modifications.cfg:
```diff
@@ -1,5 +1,35 @@
+    id=plan_unit_advance
+    name="Plan Unit Advance"
+    description="In multiplayer game, you do not control unit advances if they happen on enemy turn.
+With this add-on, you can set them beforehand:
```
These two strings without _,
and the textdomain could be specified in line 1.

add-on -> extension?

Pentarctagon replied:
Updated.

sevu commented on data/modifications/pick_advance/main.lua:
```diff
@@ -0,0 +1,249 @@
+
+wesnoth.wml_actions.set_menu_item {
+	id = "pickadvance",
+	description = "Pick Advance",
```
needs _, and due to that the textdomain needs be declared too.

This file uses _ already for something else, but I think it does not clash.

vgaming replied:
`_` is really not the best approach though, it means to mark a variable as unused in Lua... For example, for tools like `luacheck` it is the default convention and you don't get `unused variable` warnings for underscore variables. If you're really, really not interesting in an argument in Lua, you use `_`, just the underscore.

If I would use Lua best practices in Wesnoth, I'd name the function `translate` or `i18n` (or what's the better word from translation engine perspective?), e.g.:

```
local translate = wesnoth.textdomain "wesnoth"
local something = translate "Sword"
local unit = translate("Pikeman")
```

Pentarctagon replied:
The convention for Wesnoth's codebase though is to use an underscore, since that mimics how it's done with WML translations.

sevu replied:
On the Lua side in can be called anything. Using something else than _ is a matter of adding support in wmlxgettext to also recognize it.

vgaming replied:
@sevu I'd even say wmlxgettext doesn't need to be changed at all, it's a matter of local variable name `_` or `translate` or else. I guess we can close this part of the discussion as offtopic though. I should raise a separate issue for that, if I'd want to change the current Lua guidelines in wesnoth

sevu reviewed (commented):
Looks quite good, see the comments.

Wonder in which textdomain the strings should go. wesnoth-multiplayer? wesnoth-lib? wesnoth?

Pentarctagon commented:
I'll merge this in a few days barring any further comments/issues.

Pentarctagon commented on data/lua/as_text.lua:
```diff
@@ -0,0 +1,66 @@
+
+-- escaping takes 3/4 of the time, but we can't avoid it...
+local function escape(str)
+	str = string.gsub(str, delete_regexp, "")
```
Updated to `"%c"`.  Performance shouldn't be an issue here I don't think since this is only meant to be used for debugging output, and it's not exceptionally slow as-is.

Pentarctagon commented on data/lua/as_text.lua:
```diff
@@ -0,0 +1,66 @@
+local function escape(str)
+	str = string.gsub(str, delete_regexp, "")
+	str = string.gsub(str, '\\', '\\\\')
+	return string.gsub(str, '"', '\\"')
```
Updated.
